src/ModelContextProtocol.AspNetCore/McpEndpointRouteBuilderExtensions.cs

```diff
@@ -0,0 +1,108 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using ModelContextProtocol.Protocol.Messages;
+using ModelContextProtocol.Protocol.Transport;
+using ModelContextProtocol.Server;
+using ModelContextProtocol.Utils.Json;
+using System.Collections.Concurrent;
+using System.Security.Cryptography;
+
+namespace ModelContextProtocol.AspNetCore;
+
+/// <summary>
+/// Extension methods for <see cref="IEndpointRouteBuilder"/> to add MCP endpoints.
+/// </summary>
+public static class McpEndpointRouteBuilderExtensions
+{
+    /// <summary>
+    /// Sets up endpoints for handling MCP "HTTP with SSE" transport.
+    /// </summary>
+    /// <param name="endpoints">The web application to attach MCP HTTP endpoints.</param>
+    /// <param name="onSessionAsync">Provides an optional asynchronous callback for handling new MCP sessions.</param>
+    /// <returns>Returns a builder for configuring additional endpoint conventions like authorization policies.</returns>
+    public static IEndpointConventionBuilder MapMcp(this IEndpointRouteBuilder endpoints, Func<HttpContext, IMcpServer, Task>? onSessionAsync = null)
+    {
+        ConcurrentDictionary<string, SseResponseStreamTransport> _sessions = new(StringComparer.Ordinal);
+
+        var loggerFactory = endpoints.ServiceProvider.GetRequiredService<ILoggerFactory>();
+        var mcpServerOptions = endpoints.ServiceProvider.GetRequiredService<IOptions<McpServerOptions>>();
+
+        var routeGroup = endpoints.MapGroup("");
+
+        routeGroup.MapGet("/sse", async context =>
+        {
+            var response = context.Response;
+            var requestAborted = context.RequestAborted;
+
+            response.Headers.ContentType = "text/event-stream";
+            response.Headers.CacheControl = "no-cache";
+
+            var sessionId = MakeNewSessionId();
+            await using var transport = new SseResponseStreamTransport(response.Body, $"/message?sessionId={sessionId}");
+            await using var server = McpServerFactory.Create(transport, mcpServerOptions.Value, loggerFactory, endpoints.ServiceProvider);
+            _sessions.TryAdd(sessionId, transport);
+
+            try
+            {
+                var transportTask = transport.RunAsync(cancellationToken: requestAborted);
+                var serverTask = server.RunAsync(cancellationToken: requestAborted);
+                if (onSessionAsync is not null)
+                {
+                    await onSessionAsync(context, server);
+                }
+                await serverTask;
+                await transportTask;
+            }
+            catch (OperationCanceledException) when (requestAborted.IsCancellationRequested)
+            {
+                // RequestAborted always triggers when the client disconnects before a complete response body is written,
+                // but this is how SSE connections are typically closed.
```

Comment:
Really? I'd have assumed they were closed by closing the response stream / pipe. Someone actively aborts the request?

Reply:
In practice, an SSE response is canceled when the client sends a TCP FIN or an HTTP/2 RST_STREAM, both of which cancel the `HttpContext.RequestAborted` token.

Kestrel and IIS will happily allow writes to their response Streams after the TCP FIN or an HTTP/2 RST_STREAM to complete "successfully" by essentially redirecting the writes to /dev/null, because transient disconnects are so common, and too many people write important business logic like db updates in the middle of writing to the response stream. However, we observer the disconnect because of RequestAborted.

I'm super familiar with how SSE disconnects are implemented in ASP.NET Core from having made it work for SignalR in Kestrel.